client/src/rubin/nublado/client/testing/_jupyter.py

```diff
@@ -154,53 +164,89 @@ def register_extension_result(
         cache_key = normalize_source(code)
         self._extension_results[cache_key] = result
 
-    def fail(self, user: str, action: JupyterAction) -> None:
-        """Configure the given action to fail for the given user."""
-        if user not in self._fail:
-            self._fail[user] = {}
-        self._fail[user][action] = True
+    def register_python_result(self, code: str, result: str) -> None:
+        """Register the expected cell output for a given source input."""
+        self._code_results[code] = result
 
-    def _get_user_from_headers(self, request: Request) -> str | None:
-        x_user = request.headers.get("X-Auth-Request-User", None)
-        if x_user:
-            return x_user
-        # Try Authorization
-        auth = request.headers.get("Authorization", None)
-        # Is it a bearer token?
-        if auth and auth.startswith("Bearer "):
-            tok = auth[len("Bearer ") :]
-            # Is it putatively a Gafaelfawr token?
-            if tok.startswith("gt-"):
-                with suppress(Exception):
-                    # Try extracting the username. If this fails, fall through
-                    # and return None.
-                    return self._extract_user_from_mock_token(token=tok)
-        return None
+    # Below this point are the mock handler methods for the various routes.
+    # None of these methods should normally be called directly by test code.
+    # They are registered with respx and invoked automatically when a request
+    # is sent by the code under test to the mocked JupyterHub or JupyterLab.
 
     @staticmethod
-    def _extract_user_from_mock_token(token: str) -> str:
-        # remove "gt-", and split on the dot that marks the secret
-        return urlsafe_b64decode(token[3:].split(".", 1)[0]).decode()
-
-    def login(self, request: Request) -> Response:
-        user = self._get_user_from_headers(request)
-        if user is None:
-            return Response(403, request=request)
-        if JupyterAction.LOGIN in self._fail.get(user, {}):
+    def _check(
```

Comment:
Huh.  I don't think I would have thought about turning this into a decorator.  Cool.